### PR TITLE
[RFR] Heap Sort on Tiles

### PR DESCRIPTION
**Pull request by rguo123**
Working solution with tests. Assumes all the tile sorts can be loaded into memory. If not, we need to bring back two heaps

----

> [!IMPORTANT]
> Introduce heap sort for tile sorting in Deepscatter with `TileSorter` and `MinHeap`, and add tests for sorting functionality.
> 
>   - **Behavior**:
>     - Introduces `TileSorter` class in `selection.ts` for heap-based sorting of tiles.
>     - Adds `iterator()` method in `SortedDataSelection` to iterate over sorted data using `TileSorter`.
>     - Implements `MinHeap` class in `utilityFunctions.ts` for managing heap operations.
>   - **Tests**:
>     - Adds tests in `dataset.spec.js` for sorting and iterated sorting of selections using the new heap sort mechanism.
>     - Tests cover both normal and reverse iteration over sorted data.
>   - **Misc**:
>     - Minor import adjustments in `selection.ts` and `deepscatter.ts`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=nomic-ai%2Fdeepscatter&utm_source=github&utm_medium=referral)<sup> for f1ceb61b6e2cae1ef1f6771948e32d795d7aa3e4. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by rguo123** [comment]
* **#162** <a href="https://app.graphite.dev/github/pr/nomic-ai/deepscatter/162?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rguo123 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/Deeptable.ts [hunk 1]
```diff
@@ -173,6 +173,10 @@ export class Deeptable {
     });
   }
 
+  get tiles(): Tile[] {
+    return this.map(d => d);
+  }
+
```
Comment (bmschmidt) [inline src/Deeptable.ts:179]: We're going to change this to a list of tiles -- for future facing it will be important to handle undefined.

```suggestion
  get tiles(): (Tile | undefined)[] {
    return this.map(d => d);
  }

```

File: src/selection.ts [hunk 2]
```diff
@@ -1121,12 +1122,137 @@ export class SortedDataSelection extends DataSelection {
...
 
   // Given a point, returns cursor number that would select it in this selection
-  which(row: StructRowProxy) {}
+  which(row: StructRowProxy) { }
 
-  *yieldSorted(start = undefined, direction = 'up') {
-    if (start !== undefined) {
-      this.cursor = start;
+  // Q4b: what are these arguments?
+  async *iterator(start: number = 0, reverse: boolean = false) {
```
Comment (bmschmidt) [inline src/selection.ts:1128]:
```suggestion
  async *iterator(start: number = 0, reverse: boolean = false) : Iterator<StructRowProxy> {
```

**Review by bmschmidt — COMMENTED** [review]
1. Use new flatTree method with checks for undefined instead of `.tiles`

File: src/selection.ts [hunk 3]
```diff
@@ -529,7 +534,7 @@ export class DataSelection {
         // triggers creation of the deeptable column as a side-effect.
         return tile.get_column(this.name);
       }),
-    ).then(() => {});
+    ).then(() => { });
```
Comment (bmschmidt) [inline src/selection.ts:537]: Ensure that you are running `npx pretty-quick` or equivalent and no other linters on javascript

File: src/selection.ts [hunk 4]
```diff
@@ -1107,26 +1112,187 @@ export class SortedDataSelection extends DataSelection {
...
+    this.tiles = tiles;
+    this.init();
   }
 
-  // Given a point, returns cursor number that would select it in this selection
-  which(row: StructRowProxy) {}
+  // Needed to bake the tile id into the comparator.
+  getCompareFunction(): CompareFunction {
+    if (this.order === 'descending') {
+      return (a: SortInfoWithPointer, b: SortInfoWithPointer) => {
```
Comment (rguo123) [inline src/selection.ts:1167]: consider using ternary expression

File: src/selection.ts [hunk 5]
```diff
@@ -1135,28 +1301,57 @@ interface QuickSortTile {
...
+  sortIndex: number;
+  sortValue: number;
+}
+
 function quickSelect(
   k: number,
   tiles: QuickSortTile[],
   key: string,
-): StructRowProxy | undefined {
+  sortTiles: boolean = true,
```
Comment (bmschmidt) [inline src/selection.ts:1316]: Put comment about recursion here.

File: src/selection.ts [hunk 6]
```diff
@@ -1135,28 +1301,57 @@ interface QuickSortTile {
...
     }
     return undefined;
   }
 
+  // Sort the tiles to guarantee consistency
+  if (sortTiles) {
+    if (order === 'ascending') {
+      tiles.sort((a, b) => a.tile.tix - b.tile.tix);
+    } else {
+      tiles.sort((a, b) => b.tile.tix - a.tile.tix);
```
Comment (bmschmidt) [inline src/selection.ts:1351]: Again I'd use a ternary maybe with `* -1` involved

File: src/tixrixqid.ts [hunk 7]
```diff
@@ -32,6 +32,15 @@ export function zxyToTix(z: number, x: number, y: number) {
...
 }
 
+export function tileToTix(tile: Tile): number {
+  const coords = tile.key.split('/').map((d) => parseInt(d));
+  // z and x only
+  if (coords.length === 2) {
+    return zxyToTix(coords[0], coords[1], 0);
+  }
+  return zxyToTix(coords[0], coords[1], coords[2]);
+}
```
Comment (bmschmidt) [inline src/tixrixqid.ts:42]: Lose

File: tests/dataset.spec.js [hunk 8]
```diff
@@ -38,66 +36,66 @@ test('Columns can be deleted and replaced', async () => {
...
-  const selectNothing = new DataSelection(dataset, {
-    name: 'nothing and something is nothing',
-    composition: ['AND', selectThree, emptySelection],
-  });
-
-  await selectNothing.applyToAllLoadedTiles();
-  const v = selectNothing.get();
-  console.log(v);
-});
+// test('Test composition of selections', async () => {
```
Comment (bmschmidt) [inline tests/dataset.spec.js:39]: restore test shoudl work now?

File: tests/dataset.spec.js [hunk 9]
```diff
@@ -125,4 +123,116 @@ test('Test sorting of selections', async () => {
...
+  });
+  const sortKey = 'random';
+  await selectEvens.applyToAllTiles();
+  const sorted = await SortedDataSelection.fromSelection(
+    selectEvens,
+    [sortKey],
+    ({ random }) => random,
+  );
+  await sorted.applyToAllTiles();
+
```
Comment (bmschmidt) [inline tests/dataset.spec.js:141]: assert.ok(sorted.selectionSize == 4096)

File: tests/dataset.spec.js [hunk 10]
```diff
@@ -125,4 +123,116 @@ test('Test sorting of selections', async () => {
...
+  // Skip ahead 10 rows
+  for (let i = 0; i < 10; i++) {
+    first.next();
+  }
+  
+  for (let i = 0; i < 10; i++) {
+    const firstVal = first.next().value[sortKey];
+    const secondVal = second.next().value[sortKey];
+    assert.ok(Math.abs(firstVal - secondVal) < Number.EPSILON);
+  }
```
Comment (bmschmidt) [inline tests/dataset.spec.js:167]:
```suggestion

  const numbers = [0, 1, 2, 3, 4, 5];
  const firsts = numbers.map(d => firsts.next().value[sortKey])
  const firsts = numbers.map(d => second.next().value[sortKey])

  for (let i = 0; i < numbers.length; i++) {
    const firstVal = firsts[i];
    const secondVal = seconds[i];
    assert.ok(Math.abs(firstVal - secondVal) < Number.EPSILON);
  }
```

File: tests/dataset.spec.js [hunk 11]
```diff
@@ -125,4 +123,116 @@ test('Test sorting of selections', async () => {
...
+});
+
+test("Edge cases for iterated sorting of selections", async() => {
```
Comment (bmschmidt) [inline tests/dataset.spec.js:170]: Test sorting of empty selections.

**Review by bmschmidt — COMMENTED** [review]
Looks good, couple testing and merge notes

File: tests/dataset.spec.js [hunk 12]
```diff
@@ -125,4 +123,116 @@ test('Test sorting of selections', async () => {
...
 
+test('Test iterated sorting of selections', async () => {
+  const dataset = createIntegerDataset();
+  await dataset.root_tile.preprocessRootTileInfo();
+  const selectEvens = new DataSelection(dataset, {
+    name: 'twos2',
+    tileFunction: selectFunctionForFactorsOf(2),
+  });
+  const sortKey = 'random';
+  await selectEvens.applyToAllTiles();
```
Comment (bmschmidt) [inline tests/dataset.spec.js:134]:
```suggestion
```

File: tests/dataset.spec.js [hunk 13]
```diff
@@ -125,4 +123,116 @@ test('Test sorting of selections', async () => {
...
+  const sorted = await SortedDataSelection.fromSelection(
+    selectEvens,
+    [sortKey],
+    ({ random }) => random,
+  );
+  await sorted.applyToAllTiles();
```
Comment (bmschmidt) [inline tests/dataset.spec.js:140]:
```suggestion
   await sorted.applyToRootTileOnly()// <- Not a real function.
   assert.equal(sorted.selectionSize, 512 ); // <- Not the right number

  let size = 0;
  let prevValue = Number.NEGATIVE_INFINITY;
  for (const row of sorted.iterator()) {
    size++;
    // This test needs to handle that it's a structRowProxy now not a value.
    const currValue = row[sortKey];
    assert.ok(currValue >= prevValue);
    prevValue = currValue;
  }
  assert.ok(size, sorted.selectionSize);

  await selectEvens.applyToAllTiles();
```

**Review by bmschmidt — APPROVED** [review]
LGTM

**Review by bmschmidt — APPROVED** [review]
LGTM

**Comment by bmschmidt** [comment]
### Merge activity

* **Oct 24, 3:04 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/nomic-ai/deepscatter/162).
